### ISSUE TITLE
Change tf_flowers to iterate over tgz instead of extract

### ISSUE THREAD

**Issue by captain-pool** [tensorflow/datasets#658]
**Short description**
Loading tf_flowers dataset on google cloud bucket gets frozen. It works perfectly fine for local system however, passing a cloud bucket path freezes it. This is not the case with other datasets like: mnist, horses_or_humans, etc.

**Environment information**
* Operating System: Debian 9
* Python version: 3.5
* `tensorflow-datasets`/`tfds-nightly` version: 1.0.2
* `tensorflow`/`tensorflow-gpu`/`tf-nightly`/`tf-nightly-gpu` version: tensorflow==2.0.0a, tensorflow==2.0.0-beta0

**Reproduction instructions**
After logging into a Cloud VM instance, If the Path to Cloud bucket is gs://temporary_bucket
```python3
import tensorflow_datasets as tfds
ds = tfds.load("tf_flowers", manual_dir="gs://temporary_bucket")
```
**Expected behavior**
The Dataset will Load Normally, and `ds['train']` will return an `_OptionsDataset` instead of freezing forever.
**Additional context**
It works fine on Local File Systems, but, the problem is with Cloud Bucket. Other Datasets (mnist, horses_or_humans) Loads just fine.

**Comment by cyfra** [comment]
**TL;DR; slow. Would have to change tf_flowers to iterate over dataset instead of unpacking it.**

I've just checked - it does run, but very slow.

This is due to the fact, that we're unpacking whole tgz file into gs:// and it has couple thousand files. Each file creation takes a while (due to the nature of GS) and this is done single-threaded.

The solution would be to change the dataset, to iterate over the tgz file instead of unpacking it all at once. We do it already for the newer datasets.

**Comment by Suhasnama** [comment]
Hello @cyfra , @pierrot0 , @us , I would like to work up on this issue , could you please help me with this!!
Thank you!

**Comment by pierrot0** [comment]
Hi Suhasnama,

Thanks for voluntering.
You basically want to change `path = dl_manager.download_and_extract(_URL)` into `path = dl_manager.download(_URL)`, then in generate_examples, read from archive directly (`dl_manager.iter_archive(path)`).
You have an example in other datasets, for example, cats_vs_dogs: `for fname, fobj in archive:...`

**Comment by Suhasnama** [comment]
Thank yoy @pierrot0  for the help!!
Sry for the late reply.
Started working on it!!

**Comment by Suhasnama** [comment]
Hello @pierrot0 ,
I had changed `_split_generators()` to use `iterate_archive()` instead of 
`download_and_extract(_URL)` in [tf_flowers](https://github.com/Suhasnama/datasets/blob/master/tensorflow_datasets/image/flowers.py) as you mentioned.

And also changed `_generate_examples` as following , given directory structure as :
```
/flower_photos:
  /daisy:
    /1.jpg
    /2.jpg
  /dandelion:
  /roses:
  /sunflowers:
  /tulips:
```
```
def _generate_examples(self, archive):
  for fpath, fobj in archive:
          if tf.io.gfile.isdir(fpath):
              for full_path, _, fname in tf.io.gfile.walk(fpath):
                  for image in fname:
                      if image.endswith(".jpg"):
                          record = {
                              "image": os.path.join(full_path, image),
                              "label": fpath.lower(),
                          }
                          yield "%s/%s" % (fpath, image), record
```
But when I ran tests , I had encountered with
```
  File "/home/nama/git repos/datasets/tensorflow_datasets/core/download/extractor.py", line 200, in iter_archive
    return _EXTRACT_METHODS[method](path)
KeyError: <ExtractMethod.NO_EXTRACT: 1>

======================================================================
ERROR: test_download_and_prepare_as_dataset (__main__.TFFlowersTest)
test_download_and_prepare_as_dataset (__main__.TFFlowersTest)
Run the decorated test method.

```
Help me!!
Thank you.

**Comment by Conchylicultor** [comment]
@Suhasnama, just for confirmation, you still need to download the archive, right. But instead of `download_and_extract`, you have to use `download`. Have look at the examples pierrot pointed you to.

**Comment by Suhasnama** [comment]
@Conchylicultor , I am downloading the archive,
 here is the complete `_split_generators()`
```
def _split_generators(self, dl_manager):
        path = dl_manager.download(_URL)
        # There is no predefined train/val/test split for this dataset.
        return [
            tfds.core.SplitGenerator(
                name=tfds.Split.TRAIN,
                num_shards=20,
                gen_kwargs={"images_dir_path": dl_manager.iter_archive(path)},
            )
        ]
```
and here is `_generate_examples()`
```
def _generate_examples(self, images_dir_path):
    for fpath, fobj in images_dir_path:
          if tf.io.gfile.isdir(fpath):
              for full_path, _, fname in tf.io.gfile.walk(fpath):
                  for image in fname:
                      if image.endswith(".jpg"):
                          record = {
                              "image": os.path.join(full_path, image),
                              "label": fpath.lower(),
                          }
                          yield "%s/%s" % (fpath, image), record

```
and here is the directory sturcture inside archive
```
/flower_photos:
  /daisy:
    /1.jpg
    /2.jpg
  /dandelion:
  /roses:
  /sunflowers:
  /tulips:
```
Thank you!!

**Comment by Conchylicultor** [comment]
It seems it wasn't able to correctly detect the file extension. Could you print what is the `path` provided to dl_manager.iter_archive ?

**Comment by Suhasnama** [comment]
I have tried seeing `path` in `cats_vs_dogs_test.py` it points to a zip file
```
/home/nama/git repos/datasets/tensorflow_datasets/testing/test_data/fake_examples/cats_vs_dogs/cats_vs_dogs.zip
```
here(i.e, in `cats_vs_dogs.py`) it was a zip file even in tests, but in `flowers_test.py` it is not a archive in tests, but when downloading from URL
```
http://download.tensorflow.org/example_images/flower_photos.tgz
```
a archive is returned.

**Comment by Suhasnama** [comment]
@Conchylicultor  , @pierrot0  , I had submitted a PR #1055 , please suggest me changes if any.
Thank you!!

**Comment by vijayphoenix** [comment]
Taking over this issue as other PRs seems to be abandoned.
